### PR TITLE
07 data schema

### LINKED ISSUE #7: Internal data schema and validation

### PR DESCRIPTION
**Pull request by zsusswein**
This PR takes the stub codebase and replaces placeholders some initial structure. It implements

- A stub constructor workflow to demonstrate the flow of the new functionality in the class creation
- New functionality to check required inputs are provided
- New functionality to check required inputs are of appropriate types

The errors are classed and pass the calling environment to make the error messages more user-friendly. This follows [the guidance in `rlang`](https://rlang.r-lib.org/reference/topic-error-call.html). 

The eventual user flow will use `RtGam::RtGam()` as a user-facing S3 class constructor with internal helpers and validators. There will be an additional dataset constructor that does input value checking (eg., no repeated dates within groups). I originally intended to put that constructor in this PR, but I decided it made more sense to pause for feedback here in a bite-sized chunk.

In particular, feedback on (1) whether there's any required data input missing and (2) the format of the data input (strict on what's required, vector per item) will be helpful.

Closes #7

### CODE REVIEW

File: R/RtGam.R [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#' Stub: Will become fitting/class constructor
+#'
+#' Including as a stub function to showcase control flow
+#' @param cases A vector of non-negative incident case counts occurring on an
+#'   associated `reference_date`. Missing values (NAs) are not allowed.
+#' @param reference_date The associated date on which the count of incident
+#'   `cases` occurred. Missing dates are not allowed and dates can only occur
+#'   once.
+#' @param group The geographic grouping for the case/reference-date pair. Not
```
Comment (seabbs) [inline R/RtGam.R:9]: From this I take it you aren't planning on expose the `mgcv` formula interface?

File: R/checkers.R [hunk 2]
```diff
@@ -0,0 +1,148 @@
+check_required_inputs_provided <- function(cases,
+                                           reference_date,
+                                           group,
+                                           call = rlang::caller_env()) {
+  rlang::check_required(cases, "cases")
```
Comment (seabbs) [inline R/checkers.R:5]: `rlang` is a good choice here but could also use `checkmate`?

File: R/RtGam.R (hunk 1 above)
Comment (zsusswein) [reply R/RtGam.R:9]: Yeah not currently planning on exposing it. Maybe would revisit post v0.1.0?

File: R/checkers.R (hunk 2 above)
Comment (zsusswein) [reply R/checkers.R:5]: You use `checkmate` in epinowcast?

Comment (seabbs) [reply R/checkers.R:5]: no but we probably should. its used a lot in `EpiNow2`. In `epinowcast` we use a hand rolled `data.table` checker.

Comment (zsusswein) [reply R/checkers.R:5]: I like the control of the rlang builtins. Your experience on EpiNow2 was that the convenience was worth it?

Comment (zsusswein) [reply R/checkers.R:5]: I'm going to leave it with `{rlang}` here, but I like the idea in general.

File: R/RtGam.R [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#' yet implemented and a value other than `NULL` will throw an error.
+#'
+#' @returns Stub function: NULL
+#' @export
+#' @examples
+#' cases <- c(1, 2, 3)
+#' reference_date <- as.Date(c("2023-01-01", "2023-01-02", "2023-01-03"))
+#' mod <- RtGam::RtGam(cases, reference_date)
+RtGam <- function(cases,
```
Comment (seabbs) [inline R/RtGam.R:18]: Do you see a clear value in making the input be vectors vs a dataframe of cases w/ dates?

**Review by seabbs — APPROVED** [review]
This looks good to me. My only remaining question is about the decision to use an input that has a vector for cases and dates vs a dataframe or failing that no concept of dates (so you can use any time step without internal modification). 

What drove the current decision?

File: R/RtGam.R (hunk 3 above)
Comment (zsusswein) [reply R/RtGam.R:18]: > the decision to use an input that has a vector for cases and dates vs a dataframe

It's an excellent question and one I went back and forth on. I think there's a case to be made for either approach. Using a dataframe enforces for the user that the relationship between case/date/group is 1:1 and we have to construct a dataframe for mgcv anyway.

But I also find the dataframe workflow a bit clunky. It requires enforcing particular column names and these are invariably a bit different tool-to-tool. And internally, we need to break apart the dataframe into vectors anyway to do checking on each component. We can short-circuit that process by requesting vectors in the first place, which I think leads to a slightly cleaner pattern.

> failing that no concept of dates (so you can use any time step without internal modification)

This is a another good q. The motivation here is twofold:

1. I always find the date -> timestep conversion a bit tricky to get just right. We want to use small numbers (and probably z-score?) and I haven't found a super easy way to do that in R. That's the kind of thing a package with tested code is helpful with.
2.  We need to preserve that specific conversion for later prediction on arbitrary dates, so I wanted to handle it internally and get it right once.

**Comment by zsusswein** [comment]
Alright I've spent the last week going in circles on the feedback (1) around dates and (2) on formatting inputs as a dataframe vs independent vectors. Here's where I landed:

1. **Dates**: Requiring an associated datestamp for each incident case count adds an additional overhead on the user but also allows for greater flexibility. The user has to provide the additional vector of dates (or timestamps) and make sure that these line up with the case vector. It would be more convenient to only provide cases and groups. But I think this flexibility doesn't work well in the multi-group case, which is a priority of this package. If dates/timestamps aren't provided, then each group's case vector will need to be the same length, with NA padding to ensure that everything is in the right place. I think this NA padding is tricky and actually adds _more_ overhead over simply providing dates. I want to make sure multiple groups (hierarchical modeling) is supported as a first-class workflow and the preprocessing to do so is kept to a minimum.
    - For the dates vs. timestamps question, I think it's better to handle the conversion from a date to a timestamp internally rather than pushing it off to the user. As I mentioned above, doing the date to timestamp conversion can be tricky to do well and I want to handle it for the user.
2. **Dataframe vs. vector**: The expected preprocessing flow is that a user prepares a dataframe with cases, dates, and groups. This flow will be detailed in a vignette in #3. However, I still think it's worth it to have the user supply the individual columns of this dataframe as vectors (e.g., `RtGam(df[["cases"]], df[["dates"]]`) rather than the dataframe. It's convenient to avoid requiring specific column names, but I think the real benefit is the _explictness_ and _clarity_ for the users and in the code. By passing vectors, it's clear what's required, where it's supplied, and we avoid the temptation to have mutable objects or in-place changes to dataframes. I think `{EpiEstim}` benefits from this clarity -- it takes a simple incidence vector and the use is very clear. Likewise in python, `statsmodels` takes the outcome as a distinct vector from the predictors (but does allow predictors as a matrix). I like this clarity in the API and would like to keep it for now unless things become quite unwieldy.   

tl;dr I've gone back and forth, but I want keep things as-is for now. Thanks to @kaitejohnson for an outside perspective on these choices.

Thoughts @seabbs? Are you good merging as-is?

File: R/RtGam.R (hunk 3 above)
Comment (seabbs) [reply R/RtGam.R:18]: > It requires enforcing particular column names and these are invariably a bit different tool-to-tool. 

For this approach, I have been using a class constructor to prepare data. Agree it is still a little clunky but also pretty standard. 

> And internally, we need to break apart the dataframe into vectors anyway to do checking on each component. 

I'm not sure this is true if using tools for `checkmate` etc.

**Comment by seabbs** [comment]
I'm happy to see this merged and happy that a lot of thought has gone into this. 

> supply the individual columns of this dataframe as vectors (e.g., RtGam(df[["cases"]], df[["dates"]]) rather than the dataframe

However, I don't really agree with this design decision and find the arguments in favour of it quite weak. In particular, calling out `EpiEstim` as having clarity in its design language seems problematic to me as people regularly complain about this to me. My particular concern is that it locks you in to having a specific set of vectors always which means if you did want to later support a formula interface etc it would be quite hard.

**Comment by zsusswein** [comment]
In interest of keeping things moving towards a 0.1 release, I'm going to keep things as-is for now and move this conversation to an issue. I'll assign the issue to the 0.2 release.
